src/ast_editor.ml

```diff
@@ -0,0 +1,492 @@
+open Import
+
+let doc_string_uri ~document = Uri.toString (TextDocument.uri document) ()
+
+let read_html_file () =
+  let filename = Node.__dirname () ^ "/../astexplorer/dist/index.html" in
+  Fs.readFile filename
+
+let document_eq a b = Uri.equal (TextDocument.uri a) (TextDocument.uri b)
+
+let send_msg t value ~(webview : WebView.t) =
+  let msg = Ojs.empty_obj () in
+  Ojs.set_prop_ascii msg "type" (Ojs.string_to_js t);
+  Ojs.set_prop_ascii msg "value" value;
+  let (_ : bool Promise.t) = WebView.postMessage webview msg in
+  ()
+
+let fetch_pp_code ~document =
+  match Ppx_tools.get_reparsed_code_from_pp_file ~document with
+  | Ok code -> code
+  | Error err_msg ->
+    show_message `Error "%s" err_msg;
+    ""
+
+let transform_to_ast ~(document : TextDocument.t) ~(webview : WebView.t) =
+  let open Ppx_tools in
+  let origin_json =
+    Dumpast.transform
+      (TextDocument.getText document ())
+      (Pp_path.get_kind ~document)
+  in
+  let pp_value =
+    match Pp_path.get_pp_path ~document with
+    | None ->
+      show_message `Error "%s" "project root path wasn't found";
+      Jsonoo.Encode.null
```

Comment:
I find that returning `null` is suspicious. As far as I understand, this is a fatal error that signals that we can't proceed. To me, that means the right course of action is just to raise and abort the operation. Why do we need a return value at all?

Reply:
It is here because of the fact that I'm trying to send both AST (of the document, and the preprocessed one) to the Webview. If getting the preprocessed one resulted in an error, I still want to send the non preprocessed AST. 

I agree that it has to be changed. I will try to define 2 separate messages exchanges for both AST, and also only send the preprocessed one when the corresponding tab is chosen. There are also adjustments to do to update both AST on document change.